### ISSUE TITLE
Rule no-undef-init false positive

### ISSUE THREAD

**Issue by unphased** [eslint/eslint#3170]
This is an interesting situation relating to [the rule](http://eslint.org/docs/rules/no-undef-init.html).

The construction of `var v = undefined;` has a situation where it is very much not the same as `var v;`. That situation is if the code is in the body of a loop...

Now that I have been bitten by this (where we "cleaned" up (read: broke) code by heeding the suggestion of the rule) I will likely never make the same mistake again in the next year or so at least, I think that something should be done about this rule to reduce the chance that the same fate befall others.

Perhaps it is not possible to modify the rule to not fire if the declaration in question is found inside a loop construct. Then maybe a caveat can be placed in the output message.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by btmills** [comment]
The first two things I would try:

1. Move the declaration outside of the loop since it's hoisted anyway, then set it to `undefined` inside the loop, and the rule will be happy.
2. Use `// eslint-disable-line no-undef-init` on the line of the declaration to silence the rule.

**Comment by IanVS** [comment]
If you are declaring a variable within a loop and want a new variable each time, you could use `let` from ES6.  Otherwise, it's best to avoid setting `var foo` inside a loop, since as @btmills mentioned, the declaration will be hoisted out of the loop which can cause surprising bugs.

**Comment by nzakas** [comment]
@unphased as mentioned in the first comment, please provide actual source code, your rule configuration, and the actual ESLint output.

**Comment by unphased** [comment]
Well, the issue is relating to how the warning/error message prompts the end-user to potentially make a bug-creating error (as I did 4 days ago, and discovered before I posted this issue).

I don't need any help with any specific instance of code or anything. The problem is very very simple. `var x = undefined` CANNOT be harmlessly converted to `var x;` in all situations... namely, when the declaration lies inside a loop. 

Is this clear?

**Comment by nzakas** [comment]
I'm afraid this is not clear. Can you give an actual example where this is a problem? It would help to see real code and have you explain what the actual problem is. (You've only said this can't be doing safely, if like to know what bug occurs when you do.)

**Comment by michaelficarra** [comment]
@nzakas he was relying on the variable being reinitialised to undefined in each iteration of the loop.

**Comment by unphased** [comment]
Sorry I could have been more clear about it. What @michaelficarra said.

**Comment by IanVS** [comment]
@unphased can this be closed?

**Comment by unphased** [comment]
Can I suggest an adjustment (addition of a note) to the error message. Something like...

> It's not necessary to initialize '{{name}}' to undefined, *but be careful when editing such a declaration inside a loop*

**Comment by mysticatea** [comment]
Hmm, I thought this is a seed of new rule.

`no-use-uninitialized-var` -- It reports reading references before writing references.

Valid:

```js
var a = 0;
foo(a);
```

```js
var a;
if (Math.random() > 0.5) {
  a = 0;
} else {
  a = 1;
}
foo(a);
```

Invalid:

```js
var a;
if (Math.random() > 0.5) {
  a = 0;
}
foo(a);
```

This rule needs execution path analysis.

**Comment by unphased** [comment]
@mysticatea Yes that is a much more challenging thing to detect. I would imagine that detecting whether a statement is inside a loop or not would be easier, but it's also plausible that it is the same amount of difficulty. Hence my concession that perhaps we can just adjust the message.

**Comment by nzakas** [comment]
@mysticatea I think we already have too many rules that do similar things, so I'm not sure it's worthwhile to create a whole new rule.

@unphased the message is a bit too long, and I'm not sure adding that for every warning makes sense. Maybe a note of caution on the rule docs would suffice?

**Comment by unphased** [comment]
@nzakas Hmmm well... it would be better than nothing...

**Comment by gyandeeps** [comment]
I think we can safely say that we want to make doc changes here? @eslint/eslint-team

**Comment by nzakas** [comment]
Yup.

**Comment by nzakas** [comment]
Working on this.

**Comment by unphased** [comment]
Thanks!
